### ISSUE TITLE
Scoped Role Management

### ISSUE THREAD

**Issue by lgullestrup** [JosephSilber/bouncer#360]
My application is a little different from the normal 'Tenant' methodologies, since it is not the entire data structure that is shared, and multiple users are able to access multiple tenants.

I found Bouncer is a really nice and clean way to handle roles and abilities in my application.

Part of this is building a role / tenant management. 

I am trying to get a list of all the roles grouped under each specific scope for a particular use.

IE:
```
User
 | - access (Collection of tenants is indicated by scope)
        |-Tenant
             |-Roles (Collection of roles)
        |-Tenant
             |-Roles
```

I was hoping you might have some insight on how i can achieve this.

**Comment by lgullestrup** [comment]
For the time being, i am getting the roles, and the scoped separately, and then mapping the roles to the scopes.

how ever would if it was possible to get his out easily and fluently.

**Comment by JosephSilber** [comment]
I'm not very clear on what you're trying to achieve. Can you maybe reword your original question?

**Comment by lgullestrup** [comment]
Within my application, 
Users are able to cross between 'scopes' as it is identified in your library. 

IE:
```
role_id | entity_id | entity_type | scope
------------------------------------------
    1      |       1       |     User        |     1
    1      |       1       |     User        |     2
```
The nature of the organisation i am creating the app for requires this, where data needs to be separate, but together, and users need to be able to work on one set of data at a time.

I am trying to create an interface that will let a user to assigned / revoke roles against the different `scopes` available, and as such, am trying to fetch all this information in a single eloquent query that would look like the structure i drew in the first comment.

It may be that it is not possible due to the way pivot tables work.

Current solution is to create a relationship between the user and the tenants

```
public function access(){
    return $this->morphToMany(Tenant::class, 'entity', 'assigned_roles', 'entity_id', 'scope');
}
```

and then fetch all the roles as well and place the appropriate roles into the 'tenant'. but it would be nicer if it was possible to fetch it all in the once go.

Eager loading is required as the app is single page application.

I hope this clarifies my question?

**Comment by JosephSilber** [comment]
As is outlined in the [Customizing Bouncer's scope](https://github.com/JosephSilber/bouncer#customizing-bouncers-scope) section of the docs, you can create [your own `Scope` object](https://github.com/JosephSilber/bouncer/blob/master/src/Contracts/Scope.php), with its own custom implementation.

Within your `Scope`, you are free to use a `whereIn` clause to restrict the user to multiple scopes at once.
